futures-core/src/impls_nightly.rs

```diff
@@ -0,0 +1,134 @@
+use {task, Future, Stream, Poll};
+use core::mem::Pin;
+
+#[cfg(feature = "either")]
+use either::Either;
+
+impl<'a, F: ?Sized + Future> Future for &'a mut F {
+    type Output = F::Output;
+
+    fn poll(mut self: Pin<Self>, cx: &mut task::Context) -> Poll<Self::Output> {
+        unsafe { pinned_deref!(self).poll(cx) }
+    }
+}
+
+impl<'a, S: ?Sized + Stream> Stream for &'a mut S {
+    type Item = S::Item;
+
+    fn poll_next(mut self: Pin<Self>, cx: &mut task::Context) -> Poll<Option<Self::Item>> {
+        unsafe { pinned_deref!(self).poll_next(cx) }
+    }
+}
+
+impl<T: Future> Future for Option<T> {
+    type Output = Option<T::Output>;
+
+    fn poll(mut self: Pin<Self>, cx: &mut task::Context) -> Poll<Self::Output> {
+        let output = match *unsafe { Pin::get_mut(&mut self) } {
+            Some(ref mut fut) => {
+                match unsafe { Pin::new_unchecked(fut) }.poll(cx) {
+                    Poll::Ready(x) => Some(x),
+                    Poll::Pending => return Poll::Pending,
+                }
+            }
+            None => None
+        };
+        unsafe { *Pin::get_mut(&mut self) = None };
+        Poll::Ready(output)
+    }
+}
+
+if_std! {
+    use std::boxed::{Box, PinBox};
+
+    impl<'a, F: ?Sized + Future> Future for Box<F> {
+        type Output = F::Output;
+
+        fn poll(mut self: Pin<Self>, cx: &mut task::Context) -> Poll<Self::Output> {
+            unsafe { pinned_deref!(self).poll(cx) }
+        }
+    }
+
+    impl<'a, F: ?Sized + Future> Future for PinBox<F> {
+        type Output = F::Output;
+
+        fn poll(mut self: Pin<Self>, cx: &mut task::Context) -> Poll<Self::Output> {
+            unsafe {
+                let this = PinBox::get_mut(Pin::get_mut(&mut self));
+                let re_pinned = Pin::new_unchecked(this);
+                re_pinned.poll(cx)
+            }
+        }
+    }
+
+    impl<'a, F: Future> Future for ::std::panic::AssertUnwindSafe<F> {
```

Comment:
This is safe because `AssertUnwindSafe` will inherit `Unpin`ned-ness from its field (of type `F`).